### PR TITLE
Much more dynamic GHA matrix

### PR DESCRIPTION
**Pull request by mkurz**

### CODE REVIEW

File: .github/workflows/build-test.yml [hunk 1]
```diff
@@ -160,65 +160,28 @@ jobs:
...
-    needs:
-      - "extra-vars"
-      - "publish-local-jdk8"
-    uses: playframework/.github/.github/workflows/sbt.yml@v1
-    with:
-      scala: 2.13.8
-      env: SBT_VERSION=1.6.2
+      matrix: >-
+        {
+          '{0}JAVA': [ "8" ],
```
Comment (mkurz) [inline .github/workflows/build-test.yml:172]: `{0}` so we can replace it with the GHA `format` function later with `MATRIX_` or just empty string

**Comment by mkurz** [comment]
Actually I tested this quite extensively in https://github.com/mkurz/release-drafter-test/commits/master

**Comment by ihostage** [comment]
@mkurz Matthias, we already has an opportunity of dynamic matrix by using `include` logic.(https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#expanding-or-adding-matrix-configurations)
What a problem you try to solve? Maybe I just don't understand 🤷‍♂️

**Comment by ihostage** [comment]
I suggest to try something like this:
```yaml
with:
  java: >-
    [ "8" ]
  scala: >-
    [ "2.13.8" ]
  include: >-
    [ { "steps": "*1of3" }, { "steps": "*2of3" }, { "steps": "*3of3" } ]
  env: SBT_VERSION=1.6.2
```

Of course, most likely need to add feature of getting `steps` as ENV_VAR to shared workflow for using in `cmd`, but I hope you understand my idea 😄

**Comment by mkurz** [comment]
(I started to write the comment before you submitted your comment, but had to pick up child from kindergarten meanwhile)
@ihostage Right now we can not access any matrix value in the `cmd`. This is why you made the workaround with `SCALA_VERSION` [here](https://github.com/playframework/.github/blob/v1.0.4/.github/workflows/sbt-matrix.yml#L94-L95) so we can access it in `cmd` [here](https://github.com/playframework/playframework/blob/becb8ccb40f3adb5bd1b285ec70fb62cf17a5768/.github/workflows/build-test.yml#L176). The sbt version is set via `env` [here](https://github.com/playframework/playframework/blob/becb8ccb40f3adb5bd1b285ec70fb62cf17a5768/.github/workflows/build-test.yml#L171).
But in the 2.8.x build, we used to test [different sbt versions](https://github.com/playframework/playframework/blob/ca957d4a03d489643bd5d0f7853b95cdff9a55c9/.travis.yml#L14-L15). But how can we access the sbt version then? Either we have to write multiple jobs, which are basically copy-paste each time (= very redundant), only changing `env: SBT_VERSION=1.6.2` to `env: SBT_VERSION=1.3.13`. Not very nice.
So for me it makes sense to make it a matrix job. But if we do that we have to add the sbt version as env var `SBT_VERSION` like `SCALA_VERSION` [here](https://github.com/playframework/.github/blob/272efb15da484fff1544cf00d417f63a1cf9d9b3/.github/workflows/sbt-matrix.yml#L95). Same with `steps`: we have to add it as env var explict as well.

So that's why I am experimenting with this approach to not update the reusable work each time, adding a new env var when we want to add a new axis to the matrix.

But this is just an experiment right now, nothing fix yes.
I will provide the 2.8.x pull request and see how that goes.
We can talk later what we will use.

**Comment by mkurz** [comment]
Did you see https://github.com/playframework/.github/commit/0ea823aff7a204d3f883e3b9ae0b299fc472a5b9#diff-97bbbc68ac64e43b818a4f4513e417409f81d96fa38584be3526f50f874d3048R79? This is how I convert all matrix elements to env variables right now. You can not use `jq` or similiar in `${{...}}` expressions. (To use `jq` we have to make another step and use the `>> $GITHUB_ENV` approach. I can try that)

**Comment by mkurz** [comment]
I am just trying out various things. Maybe I will revert and (for 2.8.x) keep 
```yaml
with:
  java: >-
    [ "8" ]
  scala: >-
    [ "2.13.8" ]
  include: >-
    [ { "steps": "*1of3" }, { "steps": "*2of3" }, { "steps": "*3of3" }, { "sbt": "*1.6.2" }, { "sbt": "*1.3.13" } ]
```
so maybe in the end it will a combination of the existing workflow setup with results of my experiments of a bit more dynamic matrix.

File: .github/workflows/build-test.yml [hunk 2]
```diff
@@ -128,24 +115,15 @@ jobs:
...
-      java: 11
-      scala: 2.13.8
-      cmd: sbt "${{needs.extra-vars.outputs.akka_version_opts}}" "${{needs.extra-vars.outputs.akka_http_version_opts}}" ++$SCALA_VERSION Play-Integration-Test/It/test
+      matrix: >-
+        {
+          '{0}JAVA': [ "11" ],
+          '{0}SCALA': [ "2.13.8" ],
+          '{0}SBT_TEST_TASK': [ "test", "Play-Integration-Test/It/test" ],
+        }
+      cmd: sbt "${{needs.extra-vars.outputs.akka_version_opts}}" "${{needs.extra-vars.outputs.akka_http_version_opts}}" ++$MATRIX_SCALA $MATRIX_SBT_TEST_TASK
```
Comment (mkurz) [inline .github/workflows/build-test.yml:126]: @ihostage Look here. By switching to `sbt-matrix.yml` and using the new matrix axis `SBT_TEST_TASK` we save on extra job and many lines. So having the `matrix` elements as env var dynamically is nice I think, we can just always add new axis and use them as env variables.
I can also try your suggestions with `jq` or similar and keep the `-java` and `-scala` properties. We will see.

**Comment by mkurz** [comment]
This is pretty much done now. Like said, the need for more dynamic matrix jobs arose when I worked on the GHA for the 2.8.x branch.
@ihostage Like said I will see if we can keep previous the `-java` format instead of 
```yaml
      matrix: >-
        {
          '{0}JAVA': [ "8" ],
        }
```

**Comment by mkurz** [comment]
> I suggest to try something like this:
> 
> ```yaml
> with:
>   java: >-
>     [ "8" ]
>   scala: >-
>     [ "2.13.8" ]
>   include: >-
>     [ { "steps": "*1of3" }, { "steps": "*2of3" }, { "steps": "*3of3" } ]
>   env: SBT_VERSION=1.6.2
> ```

As I suspected, this does not work for our case. So we would expect following jobs (each line = one job):
* `{ java: 8, scala: 2.13.8, step: *1of3 }`
* `{ java: 8, scala: 2.13.8, step: *2of3 }`
* `{ java: 8, scala: 2.13.8, step: *3of3 }`

However, `include` does not work like that. Here, the result will be **only one job** :cry::
* `{ java: 8, scala: 2.13.8, step: *3of3 }`

Please read [this example](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategymatrixinclude) with fruit, animal,... to understand how `include` works.

Bascially that is was happening step by step:
1. The original matrix is: `{ java: 8, scala: 2.13.8 }`
1. Now GHA takes the first `include`, which has the key `step`. Does the key `step` already exist in the *original* matrix? No. So GHA will add it to each existing job: `{ java: 8, scala: 2.13.8, step: *1of3 }`. GHA will **not** create a new dimension!
1. GHA looks at the second element of the `include` , which also has the key `step`. Does the key `step` already exist in the *modified* matrix? **Yes** Now GHA will replace the element: `{ java: 8, scala: 2.13.8, step: *2of3 }`
1. Same again, the result is: `{ java: 8, scala: 2.13.8, step: *3of3 }`

I tested this here: https://github.com/mkurz/release-drafter-test/actions/runs/2251269741
As you can see I set up following matrix:
```yaml
    with:
      java: >-
         [ "11", "8" ]
      scala: >-
         [ "2.12.15", "2.13.8" ]
      include: >-
        [ { "steps": "*1of3" }, { "steps": "*2of3" }, { "steps": "*3of3" } , { "sbt_version": "1.3.13", "sbt_version": "1.6.2" } ]
```
but only 4 jobs were created. If you hover over the jobs with the mouse you see the full title:
<kbd></kbd>
So you can see alway only the last element of each `include` was added to each job: `*3of3` and `1.6.2`...
So no new dimension was created...
I already tried other things like:
```yaml
[ { "steps": [ "*1of3", "*2of3", "*3of3" ], "sbt_version": [ "1.3.13", "1.6.2" ] } ]
```
(result see [here](https://github.com/mkurz/release-drafter-test/actions/runs/2251274860))
or
```yaml
[ { "steps": [ "*1of3", "*2of3", "*3of3" ] }, { "sbt_version": [ "1.3.13", "1.6.2" ] } ]
```
(result see [here](https://github.com/mkurz/release-drafter-test/actions/runs/2251278414)) but of course that did not work too.

So there are two solution
* We use my existing soluting, where the `input` is the `matrix` string. But this is not a very nice syntax.
* We keep the `java`, `scala`, `include` and `exclude` inputs and also add a `input` like `dimensions`, where we can add more dimensions in addition to `java` and `scala`. We can then use `jq` to create one nice matrix json object, however we have to do that in a job that runs before the actual job that should use the matrix, like described here: https://stackoverflow.com/a/65434401/810109 I will try that :+1: 

BTW: I now already convert the `matrix` to env variables and that does work. See https://github.com/mkurz/release-drafter-test/blob/be9dab4a5b0fef0cc0c0950063877c677e063f79/.github/workflows/sbt-matrix.yml#L89-L91 and the result here: https://github.com/mkurz/release-drafter-test/runs/6242505670?check_suite_focus=true#step:9:9

**Comment by ihostage** [comment]
> BTW: I now already convert the matrix to env variables and that does work. See https://github.com/mkurz/release-drafter-test/blob/be9dab4a5b0fef0cc0c0950063877c677e063f79/.github/workflows/sbt-matrix.yml#L89-L91 and the result here: https://github.com/mkurz/release-drafter-test/runs/6242505670?check_suite_focus=true#step:9:9

Awesome! Thanks for work itself of my idea and brought to the end! 👍 
Now, we  just need the simple and useful syntax for matrix jobs.

>  We keep the `java`, `scala`, `include` and `exclude` inputs and also add a `input` like `dimensions`, where we can add more dimensions in addition to `java` and `scala`. We can then use `jq` to create one nice matrix json object, however we have to do that in a job that runs before the actual job that should use the matrix, like described here: https://stackoverflow.com/a/65434401/810109 I will try that 

Yep, I more like this solution, because for many repos don't need anything except `java` and `scala` and this syntax will be more understandable for other contributors. _My personal rule_: If we can do a complex work in our sharable workflow for decrease complexity in each repository and create a simple configurable workflow, we should do that. 😄 

PS: Matthias, thank you of your work 🙏 and tell me if you need my help for that 😉

**Comment by ihostage** [comment]
I really misunderstood `include` feature and I thought that this would be enough for us when I made the first version of matrix workflow. It's my mistake 😞

**Comment by mkurz** [comment]
> Now, we just need the simple and useful syntax for matrix jobs.

Actually, I am finished already. Please just see the last two commits of this pull request (and also the `dynamic_matrix` matrix branch of the reusable workflow: https://github.com/playframework/.github/compare/dynamic_matrix)
The syntax now is like:
```yaml
      java: >-
         [ "11" ]
      add-dimensions: >-
        {
          "scala": [ "2.13.8" ],
          "sbt_test_task": [ "test", "Play-Integration-Test/It/test", "Play-Microbenchmark/jmh:run -i 1 -wi 0 -f 1 -t 1 -foe=true" ],
          "folder": [ ".", "documentation" ]
        }
```
So with `add-dimensions` with can easily add new dimensions.
Also [I removed](https://github.com/playframework/.github/commit/988ca67e4cb5d8a12926dae4b03c8dc89b5fb595) the `scala` input. It is not necessary anymore. It was necessary before so we could pass `SCALA_VERSION` to the `cmd`, however now we can just use `MATRIX_SCALA` :+1:  So instead of 
```yaml
      java: >-
         [ "11" ]
      scala: >-
         [ "2.13.8" ]
```
its just adding a new dimension now (see above `add-dimensions: { "scala": [ "2.13.8" ] }`).
Also the `scala` input was required, but I think that is not a good idea, because sometimes I want to use the matrix with two java versions (8 + 11, and in future 11 + 17) but without a scala version. As an example see the `publishLocal` job of this pull request: I want to publish with Java 8 and 11, I can use `+publishLocal` which already cross publishes for all scala versions.

It is getting even better :wink: Now with `add-dimensions` and by removing the scala version, why we need `sbt.yml`? IMHO we don't need it anymore. For a simple (sbt) command you can just use
```yaml
      java: >-
         [ "11" ]
      cmd: ...
```
which will just create a one dimension matrix, so make no difference without matrix (it is the same). And if you want to add a matrix, you can just add `add-dimensions`... Very nice and flexible. That's why I propose to just remove `sbt.yml` and just maintain `sbt-matrix.yml` now.

Also to make it more comfortable I set a default Java version now: https://github.com/playframework/.github/commit/bbbd8e45cf3d7912edce2eea8e691a5cf7d66224 (like we have in `sbt.yml` already)

Last but not least: If you look at the `sbt-matrix.yml` workflow now, why actually call it `SBT-matrix.yml`? It is not really related to only a `sbt` command, you can use it for any command, so changed some sbt specific labels to be more generic: https://github.com/playframework/.github/commit/48a8368da3f939d3896e4fcf8629ab518182368e
Actually I propose to change the file name to `cmd.yml` (or `cmd-jvm-coursier-caching-matrix.yml`, but I think it is too long :wink: ) What do you think?

**Comment by mkurz** [comment]
I remove `sbt.yml` and renamed `sbt-matrix.yml` to `cmd.yml` now in https://github.com/playframework/.github/compare/dynamic_matrix
I think this is a very good solution now.

**Comment by mkurz** [comment]
Also one advantage is you can always access `MATRIX_JAVA` now when using `cmd.yml` which you coul not when using `sbt.yml` :wink:

File: .github/workflows/build-test.yml [hunk 3]
```diff
@@ -128,109 +95,53 @@ jobs:
...
       java: 11
       scala: 2.13.8
-      cmd: sbt "${{needs.extra-vars.outputs.akka_version_opts}}" "${{needs.extra-vars.outputs.akka_http_version_opts}}" ++$SCALA_VERSION test
-
-  it-tests:
-    name: Integration Tests
+      add-dimensions: >-
+        {
+          "sbt_test_task": [ "test", "Play-Integration-Test/It/test", "Play-Microbenchmark/jmh:run -i 1 -wi 0 -f 1 -t 1 -foe=true" ],
+          "folder": [ ".", "documentation" ]
```
Comment (ihostage) [inline .github/workflows/build-test.yml:105]: Can we keep jobs for docs as separated workflow? 😂
I understand that we have saved a lot of lines, but I think we loosed a readability and some contributors can be confused by this combination of `add-dimentions` and `exclude` 🤷‍♂️ 
Not everyone could comprehend GHA magic like you 😉

Comment (mkurz) [reply .github/workflows/build-test.yml:105]: I will do that 👍

**Pull request merged**